qmail.c

```diff
@@ -83,15 +93,29 @@ void qmail_to(qq,s) struct qmail *qq; char *s;
   qmail_put(qq,"",1);
 }
 
+static size_t qmail_errstr(struct qmail *qq, char *s) {
+  size_t len = 0;
+  substdio_fdbuf(&qq->ss,read,qq->fderr,qq->buf,sizeof(qq->buf));
+  while (substdio_bget(&qq->ss,s+len,1) && len < 255) {
```

Comment:
I'm not super familiar with substdio, but we do we need to read bytewise here and not all 255 bytes at once?

Also this loop won't terminate when substdio_bget returns -1 (on error).

Reply:
It does seem strange this way, and even if there’s a reason (maybe precisely because it doesn’t check for errors and assumes running out of bytes to read is the only error) I’m not sure why it needs `_bget` when there are plenty of 1-byte reads with `_get` sprinkled throughout the codebase.

Also good catch.

Reply:
On further review, the only difference between `substdio_get()` and `substdio_bget()` is that the latter might read fewer bytes at a time. This is superfluous when we’re reading a byte at a time.

For this PR, I’ll switch to one-byte `substdio_get()` with error checking to match how it’s being done everywhere else.

In a separate PR, I’ll bring back `_bget()`, mark it as deprecated, and attempt to highlight the very slight difference from `_get()`.